include/seqan3/search/algorithm/detail/search.hpp

```diff
@@ -0,0 +1,276 @@
+        max_error2.total = 0;
+        while (internal_hits.empty() && max_error2.total <= max_error.total)
+        {
+            detail::search_trivial<true>(index, query, max_error2, internal_delegate);
+            max_error2.total++;
+        }
+    }
+    else if constexpr (std::Same<remove_cvref_t<decltype(selected_mode)>, detail::search_mode_all_best>)
+    {
+        detail::search_params max_error2{max_error};
+        max_error2.total = 0;
+        while (internal_hits.empty() && max_error2.total <= max_error.total)
+        {
+            detail::search_trivial<false>(index, query, max_error2, internal_delegate);
+            max_error2.total++;
+        }
+    }
+    else if constexpr (std::Same<remove_cvref_t<decltype(selected_mode)>, strata>)
+    {
+        detail::search_params max_error2{max_error};
+        max_error2.total = 0;
+        while (internal_hits.empty() && max_error2.total <= max_error.total)
+        {
+            detail::search_trivial<true>(index, query, max_error2, internal_delegate);
+            max_error2.total++;
+        }
+        if (!internal_hits.empty())
+        {
+            internal_hits.clear(); // don't clear when using Optimum Search Schemes with lower error bounds
+            uint8_t const s = selected_mode;
+            max_error2.total += s - 1;
+            detail::search_trivial<false>(index, query, max_error2, internal_delegate);
+        }
+    }
+    else // if constexpr (std::Same<remove_cvref_t<decltype(selected_mode)>, detail::search_mode_all>)
+    {
+        detail::search_trivial<false>(index, query, max_error, internal_delegate);
+    }
+
+    // TODO: filter hits and only do it when necessary (depending on error types)
+    // _filter_hits(hits);
+
+    // output iterators or text_positions
+    if constexpr (contains<id::output_index_iterator>(cfg))
+    {
+        return internal_hits;
+    }
+    else
+    {
+        std::vector<typename index_t::size_type> hits;
+        // std::vector<std::tuple<typename index_t::size_type, detail::search_params>> hits;
+        // for (auto const & [it, error] : internal_hits)
+        auto const & selected_mode = seqan3::get<id::mode>(cfg);
+        if constexpr (std::Same<remove_cvref_t<decltype(selected_mode)>, detail::search_mode_best>)
+        {
+            // only one iterator is reported but it might contain more than one text position
+            if (!internal_hits.empty())
+            {
+                auto const & text_pos = internal_hits[0].lazy_locate();
+                hits.push_back(text_pos[0]);
+            }
+        }
+        else
+        {
+            for (auto const & it : internal_hits)
+            {
+                for (auto const & text_pos : it.locate())
+                    hits.push_back(text_pos);
+                    // hits.push_back({text_pos, error});
+                std::sort(hits.begin(), hits.end());
+                hits.erase(std::unique(hits.begin(), hits.end()), hits.end());
```

Comment:
it would be interesting to benchmark what the performance of view::unique would be compared to in-place modifications.

Reply:
That's just for the workshop. There will be better algorithms soon!